File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/map/MarkerController.java
package com.valentingrigorean.arcgis_maps_flutter.map;

import android.content.Context;

import com.esri.arcgisruntime.mapping.view.Graphic;
import com.esri.arcgisruntime.symbology.CompositeSymbol;
import com.esri.arcgisruntime.symbology.MarkerSymbol;
import com.esri.arcgisruntime.symbology.PictureMarkerSymbol;
import com.esri.arcgisruntime.symbology.Symbol;

public final class MarkerController extends BaseGraphicController implements MarkerControllerSink {
    private final CompositeSymbol marker = new CompositeSymbol();
    private final Graphic graphic;

    private BitmapDescriptor icon;
    private ScaleSymbol iconSymbol;
    private BitmapDescriptor background;
    private ScaleSymbol backgroundSymbol;

    private float opacity = 1;

    private float angle = 0.0f;

    private float iconOffsetX = 0;
    private float iconOffsetY = 0;
    private Context context;

    private boolean isSelected;

    private float selectedScale = 1.4f;


    public MarkerController(Context context, String markerId) {
        this.context = context;
        graphic = new Graphic();
        graphic.setSymbol(marker);
        graphic.getAttributes().put("markerId", markerId);
    }

    @Override
    protected Graphic getGraphic() {
        return graphic;
    }

    public Context getContext() {
        return context;
    }

    public void setSelectedScale(float selectedScale) {
        this.selectedScale = selectedScale;
        handleScaleChange();
    }

    public void setIcon(BitmapDescriptor bitmapDescriptor) {
        if (bitmapDescriptor == icon) {
            return;
        }

        if (iconSymbol != null) {
            marker.getSymbols().remove(iconSymbol);
        }
        icon = bitmapDescriptor;
        bitmapDescriptor.createSymbolAsync(new BitmapDescriptor.BitmapDescriptorListener() {
            @Override
            public void onLoaded(Symbol symbol) {
                iconSymbol = new ScaleSymbol(symbol);
                setOpacity(iconSymbol, opacity);
                setAngle(iconSymbol, angle);
                offsetSymbol(iconSymbol, iconOffsetX, iconOffsetY);
                marker.getSymbols().add(iconSymbol);
                handleScaleChange();
            }

            @Override
            public void onFailed() {

            }
        });
    }

    public void setBackground(BitmapDescriptor bitmapDescriptor) {
        if (bitmapDescriptor == background) {
            return;
        }
        if (backgroundSymbol != null) {
            marker.getSymbols().remove(backgroundSymbol);
        }

        background = bitmapDescriptor;
        bitmapDescriptor.createSymbolAsync(new BitmapDescriptor.BitmapDescriptorListener() {
            @Override
            public void onLoaded(Symbol symbol) {
                backgroundSymbol = new ScaleSymbol(symbol);

                setOpacity(backgroundSymbol, opacity);
                setAngle(backgroundSymbol, angle);
                marker.getSymbols().add(0, backgroundSymbol);
                handleScaleChange();
            }

            @Override
            public void onFailed() {

            }
        });
    }

    public void setIconOffset(float offsetX, float offsetY) {
        if (offsetX == iconOffsetX && offsetY == iconOffsetY) {
            return;
        }
        iconOffsetX = offsetX;
        iconOffsetY = offsetY;
        if (iconSymbol != null) {
            offsetSymbol(iconSymbol.getSymbol(), offsetX, offsetY);
        }
    }

    public void setOpacity(float opacity) {
        if (this.opacity == opacity) {
            return;
        }
        this.opacity = opacity;
        for (final Symbol symbol : marker.getSymbols()) {
            setOpacity(symbol, opacity);
        }
    }

    public void setAngle(float angle) {
        if (this.angle == angle) {
            return;
        }
        this.angle = angle;
        for (final Symbol symbol : marker.getSymbols()) {
            setAngle(symbol, angle);
        }
    }

    @Override
    public void setSelected(boolean selected) {
        if (selected == isSelected)
            return;
        isSelected = selected;
        handleScaleChange();
    }

    private void handleScaleChange() {
        final float scale = isSelected ? selectedScale : 1f;
        if (backgroundSymbol != null) {
            backgroundSymbol.setScale(scale);
        }
        if (iconSymbol != null) {
            iconSymbol.setScale(scale);
        }
    }

    private static void offsetSymbol(Symbol symbol, float offsetX, float offsetY) {
        if (symbol instanceof PictureMarkerSymbol) {
            final PictureMarkerSymbol pictureMarkerSymbol = (PictureMarkerSymbol) symbol;
            pictureMarkerSymbol.setOffsetX(offsetX);
            pictureMarkerSymbol.setOffsetY(offsetY);
        }
        if (symbol instanceof ScaleSymbol) {
            final ScaleSymbol scaleSymbol = (ScaleSymbol) symbol;
            offsetSymbol(scaleSymbol.getSymbol(), offsetX, offsetY);
        }
    }

    private static void setOpacity(Symbol symbol, float opacity) {
        if (symbol instanceof PictureMarkerSymbol) {
            ((PictureMarkerSymbol) symbol).setOpacity(opacity);
        }
        if (symbol instanceof ScaleSymbol) {
            final ScaleSymbol scaleSymbol = (ScaleSymbol) symbol;
            setOpacity(scaleSymbol.getSymbol(), opacity);
        }
    }

    private static void setAngle(Symbol symbol, float angle) {
        if (symbol instanceof MarkerSymbol) {
            ((MarkerSymbol) symbol).setAngle(angle);
            ((MarkerSymbol) symbol).setAngleAlignment(Float.compare(angle, 0) == 0 ? MarkerSymbol.AngleAlignment.SCREEN : MarkerSymbol.AngleAlignment.MAP);
        }
        if (symbol instanceof ScaleSymbol) {
            final ScaleSymbol scaleSymbol = (ScaleSymbol) symbol;
            setAngle(scaleSymbol.getSymbol(), angle);
        }
    }
}

File: ios/Classes/Map/MarkerController.swift
//
// Created by Valentin Grigorean on 29.03.2021.
//

import Foundation
import ArcGIS

class MarkerController: BaseGraphicController {
    private var marker: AGSCompositeSymbol
    private var icon: BitmapDescriptor?
    private var iconSymbol: ScaleSymbolHelper?
    private var backgroundImage: BitmapDescriptor?
    private var backgroundImageSymbol: ScaleSymbolHelper?

    private var iconOffsetX: CGFloat = 0
    private var iconOffsetY: CGFloat = 0

    private var opacity: Float = 1

    private var selected: Bool = false

    private var selectedScale: CGFloat = 1.4

    private var angle: Float = 0.0

    init(markerId: String,
         selectionPropertiesHandler: SelectionPropertiesHandler) {
        marker = AGSCompositeSymbol()
        super.init(graphics: AGSGraphic(geometry: nil, symbol: marker, attributes: ["markerId": markerId]), selectionPropertiesHandler: selectionPropertiesHandler)
    }

    override var isSelected: Bool {
        get {
            selected
        }
        set {
            if selected != newValue {
                selected = newValue
                handleScaleChanged()
            }
        }
    }

    func setIcon(bitmapDescription: BitmapDescriptor) {
        if icon == bitmapDescription {
            return
        }
        if icon != nil {
            marker.symbols.remove(at: backgroundImage == nil ? 0 : 1)
        }
        icon = bitmapDescription
        iconSymbol = ScaleSymbolHelper(symbol: createSymbol(bitmapDescription: bitmapDescription))
        offsetSymbol(symbol: iconSymbol!.symbol, offsetX: iconOffsetX, offsetY: iconOffsetY)
        marker.symbols.insert(iconSymbol!.symbol, at: backgroundImage == nil ? 0 : 1)
        updateCommonProps()
    }

    func setBackground(bitmapDescription: BitmapDescriptor) {
        if backgroundImage == bitmapDescription {
            return
        }
        if backgroundImage != nil {
            marker.symbols.remove(at: 0)
        }

        backgroundImage = bitmapDescription
        backgroundImageSymbol = ScaleSymbolHelper(symbol: createSymbol(bitmapDescription: bitmapDescription))
        marker.symbols.insert(backgroundImageSymbol!.symbol, at: 0)
        updateCommonProps()
    }


    func setIconOffset(offsetX: CGFloat,
                       offsetY: CGFloat) {
        if offsetX == iconOffsetX && offsetY == iconOffsetY {
            return
        }
        iconOffsetX = offsetX
        iconOffsetY = offsetY

        if icon != nil {
            guard  let symbol = marker.symbols.last as? AGSMarkerSymbol else {
                return
            }
            offsetSymbol(symbol: symbol, offsetX: offsetX, offsetY: offsetY)
        }
    }

    func setOpacity(opacity: Float) {
        self.opacity = opacity
        setGraphicsOpacity(opacity: opacity)
    }

    func setAngle(angle: Float) {
        if self.angle == angle {
            return
        }
        self.angle = angle
        setGraphicsAngle(angle: angle)
    }

    func setSelectedScale(selectedScale: CGFloat) {
        if selectedScale == selectedScale {
            return
        }
        self.selectedScale = selectedScale
        handleScaleChanged()
    }

    private func updateCommonProps() {
        handleScaleChanged()
        setGraphicsAngle(angle: angle)
        setGraphicsOpacity(opacity: opacity)
    }

    private func handleScaleChanged() {

        let scale = isSelected ? selectedScale : 1.0

        if let iconSymbol = iconSymbol {
            iconSymbol.setScale(scale: scale)
        }

        if let backgroundImageSymbol = backgroundImageSymbol {
            backgroundImageSymbol.setScale(scale: scale)
        }
    }

    private func offsetSymbol(symbol: AGSSymbol,
                              offsetX: CGFloat,
                              offsetY: CGFloat) {
        guard let markerSymbol = symbol as? AGSMarkerSymbol else {
            return
        }

        markerSymbol.offsetX = offsetX
        markerSymbol.offsetY = offsetY
    }

    private func setGraphicsOpacity(opacity: Float) {
        for symbol in marker.symbols {
            if let markerSymbol = symbol as? AGSPictureMarkerSymbol {
                markerSymbol.opacity = opacity
            }
        }
    }

    private func setGraphicsAngle(angle: Float) {
        for symbol in marker.symbols {
            if let markerSymbol = symbol as? AGSMarkerSymbol {
                markerSymbol.angle = angle
            }
        }
    }

    private func createSymbol(bitmapDescription: BitmapDescriptor) -> AGSSymbol {
        let newSymbol = bitmapDescription.createSymbol()
        setOpacity(opacity: opacity)
        return newSymbol
    }

    private func setOpacity(symbol: AGSSymbol,
                            opacity: Float) {
        guard let pictureSymbol = symbol as? AGSPictureMarkerSymbol else {
            return
        }
        pictureSymbol.opacity = opacity
    }
}
